### PR TITLE
Don't apply root padding to nested layout blocks

### LINKED ISSUE #43095: Root padding is doubled on child blocks
### Description

A lot of effort has gone into consolidating and simplifying block theme support for making blocks fullwidth, while also supporting root padding, particularly in https://github.com/WordPress/gutenberg/pull/42085. 

I've been testing this out with TT2 and it looks like padding is inaccurately applied to child blocks, doubling up the padding within. 

Do we need the following generated styles — or should we be zeroing out padding if `.has-global-padding ` > `.has-global-padding`  occurs? 

```
.has-global-padding > .alignfull > :where([class*="wp-block-"]:not(.alignfull):not(.alignfull):not([class*="__"]),p,h1,h2,h3,h4,h5,h6,ul,ol) {
    padding-right: var(--wp--style--root--padding-right);
    padding-left: var(--wp--style--root--padding-left);
}
```

I think we're really close on getting a root padding/fullwidth that "just works" 🤞

### Step-by-step reproduction instructions

1. Start with the TT2 theme.
1. Comment out/remove all the custom "Alignment styles" from the theme ([source](https://github.com/WordPress/twentytwentytwo/blob/trunk/style.css#L86-L134)).
2. Add `"settings.useRootPaddingAwareAlignments": true` to the theme.json
3. Update the theme.json styles.spacing.padding values as such, which emulates what TT2 was achieving with custom alignment styles: 
```
"padding": {
       "top": "0px",
        "right": "var(--wp--custom--spacing--outer)",
       "bottom": "0px",
        "left": "var(--wp--custom--spacing--outer)"
} 
```
4. Add the TT2 "Video with header and details" pattern.
5. View the front-end/editor, resizing the viewport to see the issue (as double the proper padding is applied). 

### Screenshots, screen recording, code snippet

`.has-global-padding` is being applied to subsequent fullwidth blocks, which doubles the amount of left/right padding. This doubled padding is only perceptible on mobile as the viewport is small enough to see the effect:

Frame 3

Frame 2

### Environment info

WordPress 6.0.1
Twenty Twenty Two (modified from above)

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by tellthemachines**
## What?

Fixes #43095 by resetting each root padding rule so it doesn't apply where there are nested `.has-global-padding` selectors. The new rules added simply negate the existing rules if the `.has-global-padding` block is nested inside another `.has-global-padding` block.

## Behaviour before fix

* Add left/right root padding value to all blocks with `.has-global-padding`;
* Set negative left/right margin with same value on all `.alignfull` direct children of blocks with `.has-global-padding`;
* Set left/right root padding value on all blocks that are direct children of the `.alignfull` blocks with negative margins.

## Expected behaviour with fix applied

1. The outermost container block with content width set will get left/right root padding value;
2. Its direct `alignfull` children will get negative left/right margins;
3. The direct children of those alignfull children will also get left/right root padding value.
4. Any nested blocks with content width set will _not_ get left/right root padding value _unless_ they are also direct children of the `alignfull`blocks with negative margins mentioned in 2..

I've tested this with a bunch of different nesting patterns and it seems to be working as expected. The one place where there can be a bit of weirdness is in the post editor view, because the `.has-global-padding` class isn't output at root level even if the Post Content block has layout set. This means that a block with layout in the post editor might display some extra padding:
Screen Shot 2022-09-05 at 3 54 47 pm
which won't appear in the front end view:
Screen Shot 2022-09-05 at 3 55 10 pm

I think the only way to fully fix that is to make it possible for the post editor to be aware of the Post Content block settings, which we should look into doing separately from this fix.

## Testing Instructions

1. In theme.json, add `"useRootPaddingAwareAlignments": true` inside the settings object;
2. In both post and site editor, add a few container blocks (e.g. Group) nested inside each other, with other content inside them. Try with both full width and regular width set, and try adding full width content (e.g. Image block) inside them too;
3. Check that root padding behaves as expected.

## Screenshots or screencast 

Before:
Screen Shot 2022-09-05 at 4 30 14 pm

After:
Screen Shot 2022-09-05 at 4 29 21 pm

### CODE REVIEW

File: lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
Comment (richtabor) [inline lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: Shouldn't this also have `:not(.alignwide)` included as well? Since the `.has-global-padding` class is already applied to the parent block? 

Currently `wide` width blocks inside a group block with global padding still get 2x the proper padding. 

CleanShot 2022-09-05 at 20 34 47@2x

Comment (richtabor) [reply lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: Seems like the remaining issue occurs when a wide width block is within a fullwidth block (that's not omitted from the above code).

Comment (richtabor) [reply lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: It's especially noticeable when you add padding to the parent group block, regardless of constrained layout or not. 

Editor: 

CleanShot 2022-09-05 at 20 45 55@2x

Front-end: 

CleanShot 2022-09-05 at 20 46 21@2x

You can try [this snippet](https://gist.github.com/richtabor/19ceab88f73b6ad40c58d977879da13b), with Twenty twenty Three to replicate.

Comment (richtabor) [reply lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: Thanks for digging in on this — I know its a tricky one.

Comment (tellthemachines) [reply lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: Yeah, that padding will be applied to any block inside the nested `alignfull` block that's not also `alignfull`; you can see it also gets applied to a regular width block:
Screen Shot 2022-09-06 at 5 48 20 pm

(the yellow Group blocks are outside the full width Group and the Media&Text ones inside it)

This is also complicated by the `box-sizing` property on the blocks that receive padding: some of them are set to `border-box` (e.g. Media&Text) and others to `content-box` (e.g. Paragraph, Image) so results vary depending on the block types.

I'm thinking we could target this extra padding rule only to children of `alignfull` blocks that _don't_ have content width set, as they're the ones that really need that padding? Whereas in blocks with content width set it's more likely to break the alignment.

Comment (tellthemachines) [reply lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: Ok I've tried ignoring children of `alignfull` blocks with content width.

File: packages/edit-site/src/components/global-styles/use-global-styles-output.js
Comment (andrewserong) [inline packages/edit-site/src/components/global-styles/use-global-styles-output.js]: This line appears to break global styles output for me. Can we split this into multiple lines so that it's easier to read / work out where we're concatenating `}` characters? (I think one of the rules is missing a closing brace, and the very final rule has an extra closing brace since the below `ruleset += '}';` line will add a closing one)

image

File: lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
Comment (andrewserong) [inline lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: Not necessarily a blocker, but now that these rules are fairly complex, is it worth adding a comment before each line about what the rule does? The other rules in this function feel fairly self explanatory to me, but just thinking of our future selves if we want to update these rules at any point, so that we don't forget why each line exists.

**Review by andrewserong — COMMENTED** [review]
This is testing pretty well for me @tellthemachines! With @richtabor's provided markup, when I switch the Media & Text block from wide alignment to full alignment, the nested full width alignment block no longer has extra padding applied 👍

Just left a couple of comments — now that the rules are fairly complex, it'd be good to try to add some inline comments explaining each rule, to aid future debugging. The only blocker I ran into was with an issue with the output in global styles, but looks like it should be a fairly simple fix 🤞

File: packages/edit-site/src/components/global-styles/use-global-styles-output.js
Comment (tellthemachines) [reply packages/edit-site/src/components/global-styles/use-global-styles-output.js]: Ooooh well spotted, yeah this needs to be more legible 😅

Comment (tellthemachines) [reply packages/edit-site/src/components/global-styles/use-global-styles-output.js]: Done!

**Comment by tellthemachines** [comment]
Thanks for the review @andrewserong ! I included some inline comments on what those CSS rules do. Hopefully they're clear enough! I'll also be working on a dev note for this functionality next week, that will further document the rules and our assumptions about how they should work. Once the feature's in Core we should add that documentation to the handbook.

**Review by andrewserong — APPROVED** [review]
Thanks @tellthemachines, those added comments are super helpful and make it much easier to read 👍

This appears to be testing well for me in the site editor as well as the post editor now. It might be worth getting another pair of eyes from @richtabor or others to confirm that the fix largely addresses the pain points. But, given the RC for the next Gutenberg release isn't until next Tuesday, we could always merge and follow-up separately with any required tweaks, too.

LGTM! 🎉

**Comment by tellthemachines** [comment]
I'll go ahead and merge this; if anything's still off we can address it in a follow-up.

**Pull request merged**
